nixos/modules/services/mail/dovecot.nix

```
@@ -105,107 +109,205 @@ let
     }) cfg.sieve.pipeBins
   );
 
-  dovecotConf = concatStrings [
-    ''
-      base_dir = ${baseDir}
-      protocols = ${concatStringsSep " " cfg.protocols}
-      sendmail_path = /run/wrappers/bin/sendmail
-      mail_plugin_dir = /run/current-system/sw/lib/dovecot/modules
-      # defining mail_plugins must be done before the first protocol {} filter because of https://doc.dovecot.org/configuration_manual/config_file/config_file_syntax/#variable-expansion
-      mail_plugins = $mail_plugins ${concatStringsSep " " cfg.mailPlugins.globally.enable}
-    ''
+  dovecotConf =
+    if versionAtLeast cfg.package.version "2.4" then
+      concatStrings [
 
-    (concatStringsSep "\n" (
-      mapAttrsToList (protocol: plugins: ''
-        protocol ${protocol} {
-          mail_plugins = $mail_plugins ${concatStringsSep " " plugins.enable}
-        }
-      '') cfg.mailPlugins.perProtocol
-    ))
+        ''
+          dovecot_config_version = ${cfg.settings.configVersion}
+          dovecot_storage_version = ${cfg.settings.storageVersion}
+          base_dir = ${baseDir}
+          protocols = ${concatStringsSep " " cfg.protocols}
+          sendmail_path = /run/wrappers/bin/sendmail
+          mail_plugin_dir = /run/current-system/sw/lib/dovecot/modules
+          # defining mail_plugins must be done before the first protocol {} filter because of https://doc.dovecot.org/configuration_manual/config_file/config_file_syntax/#variable-expansion
```

The link is stale, for 2.4 the link should be
```suggestion
          # defining mail_plugins must be done before the first protocol {} filter because of https://doc.dovecot.org/2.4.0/core/settings/syntax.html#variable-expansion
```